dexdcr/wallet.go

```diff
@@ -259,55 +259,64 @@ func (spvw *SpvWallet) LockUnspent(ctx context.Context, unlock bool, ops []*wire
-	// Attempt to read the unspent txout info from wallet.
-	// gettxout in spv mode returns details for outputs as long
-	// they exist, pay to the wallet and are unspent.
-	outpoint := wire.OutPoint{Hash: *txHash, Index: index, Tree: tree}
-	utxo, err := spvw.wallet.UnspentOutput(ctx, outpoint, mempool)
-	if err != nil {
-		if errors.Is(err, errors.NotExist) {
-			return nil, asset.CoinNotFoundError
-		}
-		return nil, err
+func (spvw *SpvWallet) UnspentOutput(ctx context.Context, txHash *chainhash.Hash, index uint32, tree int8) (*dcr.TxOutput, error) {
+	// Attempt to read the unspent txout info from wallet. The output
+	// must exist, pay to the wallet and be unspent.
+	var checkTrees []int8
+	switch {
+	case tree == wire.TxTreeUnknown:
+		checkTrees = []int8{wire.TxTreeRegular, wire.TxTreeStake}
+	case tree == wire.TxTreeRegular || tree == wire.TxTreeStake:
+		checkTrees = []int8{tree}
+	default:
+		return nil, fmt.Errorf("invalid tx tree %d", tree)
 	}
 
-	// Disassemble script into single line printable format.  The
-	// disassembled string will contain [error] inline if the script
-	// doesn't fully parse, so ignore the error here.
-	disbuf, _ := txscript.DisasmString(utxo.PkScript)
+	for _, tree = range checkTrees {
+		outpoint := wire.OutPoint{Hash: *txHash, Index: index, Tree: tree}
+		utxo, err := spvw.wallet.UnspentOutput(ctx, outpoint, true)
+		if err != nil {
+			if errors.Is(err, errors.NotExist) {
+				continue // check next tree
+			}
+			return nil, err
+		}
 
-	// Get further info about the script.  Ignore the error here since an
-	// error means the script couldn't parse and there is no additional
-	// information about it anyways.
-	scriptClass, addrs, reqSigs, _ := txscript.ExtractPkScriptAddrs(
-		0, utxo.PkScript, spvw.chainParams, true) // Yes treasury
-	addresses := make([]string, len(addrs))
-	for i, addr := range addrs {
-		addresses[i] = addr.String()
-	}
+		// Get further info about the script.  Ignore the error here since an
+		// error means the script couldn't parse and there is no additional
+		// information about it anyways.
+		_, addrs, _, _ := txscript.ExtractPkScriptAddrs(
+			0, utxo.PkScript, spvw.chainParams, true) // Yes treasury
+		addresses := make([]string, len(addrs))
+		for i, addr := range addrs {
+			addresses[i] = addr.String()
+		}
+
+		_, bestHeight := spvw.wallet.MainChainTip(ctx)
+		var confirmations uint32
+		if utxo.Block.Height != -1 {
+			confirmations = uint32(confirms(utxo.Block.Height, bestHeight))
+		}
 
-	bestHash, bestHeight := spvw.wallet.MainChainTip(ctx)
-	var confirmations int64
-	if utxo.Block.Height != -1 {
-		confirmations = int64(confirms(utxo.Block.Height, bestHeight))
+		return &dcr.TxOutput{
+			TxOut: &wire.TxOut{
+				Value:    int64(utxo.Amount),
+				PkScript: utxo.PkScript,
+				// Version: 0, // TODO: dcrwallet json-rpc doesn't set this also, but should be
```

Comment:
This would be done in another PR right?

Reply:
It'll be done when dcrwallet does it too. The idea is to closely simulate the dcrwallet json-rpc methods as that's what the main dex project uses for dcr wallet operations. It's something that should ideally be done but not sure there are any plans towards that yet. It's safe as is for now though.